File: kaiwu/tests/bench_tasks/t44_go_cache/cache.go
// Package cache implements a thread-safe LRU cache with TTL support.
// Bugs:
// 1. cache.go: Get() does not move accessed entry to front (breaks LRU order)
// 2. cache.go: evict() removes from front but LRU should be at back (list direction wrong)
// 3. stats.go: HitRate() divides by Hits instead of total requests
package cache

import (
	"container/list"
	"sync"
	"time"
)

type entry struct {
	key       string
	value     interface{}
	expiresAt time.Time
}

// Cache is a thread-safe LRU cache with optional TTL.
type Cache struct {
	mu       sync.Mutex
	cap      int
	items    map[string]*list.Element
	order    *list.List // front = most recently used
	stats    Stats
	defaultTTL time.Duration
}

// New creates a new Cache with the given capacity and default TTL.
// TTL of 0 means no expiry.
func New(capacity int, defaultTTL time.Duration) *Cache {
	return &Cache{
		cap:        capacity,
		items:      make(map[string]*list.Element),
		order:      list.New(),
		defaultTTL: defaultTTL,
	}
}

// Set stores a key-value pair. Uses defaultTTL if ttl is 0.
func (c *Cache) Set(key string, value interface{}, ttl time.Duration) {
	c.mu.Lock()
	defer c.mu.Unlock()

	if ttl == 0 {
		ttl = c.defaultTTL
	}

	var exp time.Time
	if ttl > 0 {
		exp = time.Now().Add(ttl)
	}

	if el, ok := c.items[key]; ok {
		c.order.MoveToFront(el)
		el.Value.(*entry).value = value
		el.Value.(*entry).expiresAt = exp
		return
	}

	el := c.order.PushFront(&entry{key: key, value: value, expiresAt: exp})
	c.items[key] = el

	if c.order.Len() > c.cap {
		c.evict()
	}
}

// Get retrieves a value by key. Returns (value, true) or (nil, false).
func (c *Cache) Get(key string) (interface{}, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()

	el, ok := c.items[key]
	if !ok {
		c.stats.Misses++
		return nil, false
	}

	e := el.Value.(*entry)
	if !e.expiresAt.IsZero() && time.Now().After(e.expiresAt) {
		c.removeElement(el)
		c.stats.Misses++
		return nil, false
	}

	// Bug: does not move to front on access (breaks LRU ordering)
	c.stats.Hits++
	return e.value, true
}

// Delete removes a key from the cache.
func (c *Cache) Delete(key string) bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	el, ok := c.items[key]
	if !ok {
		return false
	}
	c.removeElement(el)
	return true
}

func (c *Cache) removeElement(el *list.Element) {
	c.order.Remove(el)
	delete(c.items, el.Value.(*entry).key)
}

// evict removes the least recently used entry.
// LRU is at the back of the list (front = MRU).
func (c *Cache) evict() {
	// Bug: removes from front (MRU) instead of back (LRU)
	el := c.order.Front()
	if el != nil {
		c.removeElement(el)
		c.stats.Evictions++
	}
}

// Len returns the number of items in the cache.
func (c *Cache) Len() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.order.Len()
}

// Stats returns a copy of the cache statistics.
func (c *Cache) Stats() Stats {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.stats
}

File: kaiwu/tests/bench_tasks/t34_cache_invalidation/cache.py
"""Cache store with TTL and LRU eviction."""

import time
from collections import OrderedDict
from typing import Any, Optional


class CacheEntry:
    def __init__(self, value: Any, ttl: float, created_at: float):
        self.value = value
        self.ttl = ttl
        self.created_at = created_at

    def is_expired(self, now: float) -> bool:
        if self.ttl <= 0:
            return False  # no expiry
        return now - self.created_at >= self.ttl


class LRUCache:
    """LRU cache with optional per-entry TTL."""

    def __init__(self, max_size: int, default_ttl: float = 0):
        self.max_size = max_size
        self.default_ttl = default_ttl
        # OrderedDict: most-recently-used at end
        self._store: OrderedDict[str, CacheEntry] = OrderedDict()
        self.hits = 0
        self.misses = 0
        self.evictions = 0

    def get(self, key: str, now: float = None) -> Optional[Any]:
        if now is None:
            now = time.monotonic()
        if key not in self._store:
            self.misses += 1
            return None
        entry = self._store[key]
        if entry.is_expired(now):
            # Bug: does not remove the expired entry from the store
            self.misses += 1
            return None
        # Move to end (most recently used)
        self._store.move_to_end(key)
        self.hits += 1
        return entry.value

    def set(self, key: str, value: Any, ttl: float = None, now: float = None) -> None:
        if now is None:
            now = time.monotonic()
        effective_ttl = ttl if ttl is not None else self.default_ttl
        if key in self._store:
            self._store.move_to_end(key)
        self._store[key] = CacheEntry(value, effective_ttl, now)
        if len(self._store) > self.max_size:
            self._evict_lru()

    def _evict_lru(self) -> None:
        """Remove the least-recently-used entry."""
        self._store.popitem(last=False)
        self.evictions += 1

    def delete(self, key: str) -> bool:
        if key in self._store:
            del self._store[key]
            return True
        return False

    def clear(self) -> None:
        self._store.clear()

    def size(self) -> int:
        return len(self._store)

    def keys(self) -> list[str]:
        return list(self._store.keys())
